src/main/java/io/github/thebusybiscuit/slimefun4/core/services/JsonDeserializationService.java

```diff
@@ -0,0 +1,160 @@
+package io.github.thebusybiscuit.slimefun4.core.services;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+import me.mrCookieSlime.Slimefun.api.Slimefun;
+import org.bukkit.configuration.serialization.ConfigurationSerialization;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.logging.Level;
+
+/**
+ * Represents a mapper between classes and json deserialization logic. This class is
+ * somewhat similar to Bukkit's {@link ConfigurationSerialization} class.
+ *
+ * @author md5sha256
+ */
+public class JsonDeserializationService {
+
+    public static final String ENUM_CLASS_KEY = "enum-class", ENUM_VALUE_KEY = "value";
+
+    private final JsonParser parser = new JsonParser();
+    private final Map<String, Function<JsonElement, ?>> deserializerMap = new HashMap<>();
+
+    public JsonDeserializationService() {
+    }
+
+    /**
+     * Convenience method to deserialize an {@link Enum} which was serialized via {@link #serializeEnum(Enum)}
+     *
+     * @param element The {@link JsonElement} to deserialize
+     * @return Returns the deserialize enum instance or null
+     * @see #serializeEnum(Enum)
+     */
+    @Nullable
+    public static Enum<?> deserializeEnum(@Nonnull final JsonElement element) {
+        if (!element.isJsonObject()) {
+            return null;
+        }
+        final JsonObject object = element.getAsJsonObject();
+        final JsonElement clazz = object.get(ENUM_CLASS_KEY), value = object.get(ENUM_VALUE_KEY);
+        final JsonPrimitive clazzString, enumValue;
```

Comment:
Formatting :NotLikeThis:

```suggestion
        JsonObject object = element.getAsJsonObject();
        JsonElement clazz = object.get(ENUM_CLASS_KEY);
        JsonElement value = object.get(ENUM_VALUE_KEY);
        JsonPrimitive clazzString;
        JsonPrimitive enumValue;
```